vmm/src/memory_manager.rs

```diff
@@ -1300,30 +1299,42 @@ impl MemoryManager {
         host_numa_node: Option<u32>,
         existing_memory_file: Option<File>,
     ) -> Result<Arc<GuestRegionMmap>, Error> {
-        let (f, f_off) = if let Some(f) = existing_memory_file {
-            (f, file_offset)
+        let mut mmap_flags = libc::MAP_NORESERVE;
+
+        // The duplication of mmap_flags ORing here is unfortunate but it also makes
+        // the complexity of the handling clear.
+        let fo = if let Some(f) = existing_memory_file {
+            if shared {
+                mmap_flags |= libc::MAP_SHARED;
+            } else {
+                mmap_flags |= libc::MAP_PRIVATE;
+            }
+            Some(FileOffset::new(f, file_offset))
+        } else if let Some(backing_file) = backing_file {
+            if shared {
+                mmap_flags |= libc::MAP_SHARED;
+            } else {
+                mmap_flags |= libc::MAP_PRIVATE;
+            }
+            Some(Self::open_backing_file(backing_file, file_offset, size)?)
+        } else if shared || hugepages {
+            // For hugepages we must also MAP_SHARED otherwise we will trigger #4805
+            // because the MAP_PRIVATE will trigger CoW against the backing file with
+            // the VFIO pinning
+            mmap_flags |= libc::MAP_SHARED;
```

Comment:
I think we should error out if the user specified `shared=off` and `hugepages=on`.
This is silently overriding `shared=off` which could be rather confusing.

Reply:
I think we should warn and coerce the value in the config validation.